src/network/serverpackethandler.cpp

```diff
@@ -1346,20 +1348,37 @@ void Server::handleCommand_RemovedSounds(NetworkPacket* pkt)
 	}
 }
 
+static void sanitize_field(std::string &s)
+{
+	// truncate on NULL
+	auto pos = s.find('\000');
+	if (pos != s.npos)
+		s.erase(pos);
+
+	// remove control characters except tab/feed
+	s.erase(std::remove_if(s.begin(), s.end(), [] (unsigned char c) -> bool {
+		return c < 9 || (c >= 13 && c < 32);
+	}), s.end());
+}
+
 static bool pkt_read_formspec_fields(NetworkPacket *pkt, StringMap &fields)
 {
 	u16 field_count;
 	*pkt >> field_count;
 
-	u64 length = 0;
+	size_t length = 0;
 	for (u16 k = 0; k < field_count; k++) {
-		std::string fieldname;
+		std::string fieldname, fieldvalue;
 		*pkt >> fieldname;
-		fields[fieldname] = pkt->readLongString();
+		fieldvalue = pkt->readLongString();
+		sanitize_field(fieldname);
+		sanitize_field(fieldvalue);
+
+		length += fieldname.size() + fieldvalue.size();
 
-		length += fieldname.size();
-		length += fields[fieldname].size();
+		fields[fieldname] = std::move(fieldvalue);
```

Comment:
```suggestion
		fields[std::move(fieldname)] = std::move(fieldvalue);
```

(though probably irrelevant given how short names can be expected to be)

Reply:
ah huh, I thought `operator[]` didn't work with xvalues